### ISSUE TITLE
Alluxio Master memory leak

### ISSUE THREAD

**Issue by jingcheng88** [Alluxio/alluxio#14746]
**Alluxio Version:**
2.6.1 2.6.2 2.7.1

**Describe the bug**
There are 30 worker in the cluster. The memory of master continued to grow and OOM in the end. The memory did not GC even if there were no client request. The GC of JVM was well. 

![image](https://user-images.githubusercontent.com/8472283/147311733-fef02b1a-5aab-47e2-b655-6a10263bbf38.png)

![image](https://user-images.githubusercontent.com/8472283/147311593-369d447a-f136-4e54-b248-98b2e4ee3c12.png)

**Comment by jiacheliu3** [comment]
Hi @jingcheng88 Did you look into where the memory consumption is from? Do you have jstacks or heap dumps or anything else that we can investigate like `jmap -histo <pid>`?

**Comment by jingcheng88** [comment]
@jiacheliu3
[alluxio_stack.txt](https://github.com/Alluxio/alluxio/files/7779844/alluxio_stack.txt)
[jmap.txt](https://github.com/Alluxio/alluxio/files/7779849/jmap.txt)
```
$ /usr/lib/jvm/java-1.8-openjdk/bin/jstat -gc 6 5000 1000
 S0C    S1C    S0U    S1U      EC       EU        OC         OU       MC     MU    CCSC   CCSU   YGC     YGCT    FGC    FGCT     GCT
287232.0 343552.0 283000.1  0.0   4904960.0 1094161.7 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1099648.0 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1101438.9 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1102825.9 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1105130.4 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1110210.5 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1116152.3 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1118164.0 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1139233.6 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1184163.0 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
287232.0 343552.0 283000.1  0.0   4904960.0 1185160.5 11185152.0 1641350.6  88832.0 83862.1 10752.0 9812.5     22    2.837   0      0.000    2.837
```

**Comment by jiacheliu3** [comment]
@jingcheng88 What JVM are you using? The jstack cannot be parsed by fastthread.io. The format looks different from my jstack from openjdk. It's a bit hard to do automated analysis, so I'll come back to this in a bit.

Your jmap output reveals something interesting, let's go over them one by one.

First of all, your heap looks very small. That means some constant-cost objects will rank high in `jmap`. When troubleshooting something like a leak, we should ignore those constant-cost objects. 

```
 num     #instances         #bytes  class name
----------------------------------------------
   1:      18352143      440M  java.lang.Long
   2:        224637      346M  [I
   3:       2539433      270M  [B
   4:       2781776      228M  [C
   5:       2036152      195M  [Ljava.lang.Object;
   6:       3251304      130M  java.lang.ref.Finalizer
   7:       2918450       93M  org.apache.ratis.server.protocol.TermIndex$1
   8:           980       84M  [Ljava.util.concurrent.ConcurrentHashMap$Node;
   9:        471498       75M  alluxio.master.file.meta.MutableInodeFile
  10:       2630772       63M  java.lang.String
  11:       1038307       58M  org.apache.ratis.proto.RaftProtos$LogEntryProto
  12:       1573735       50M  java.util.concurrent.ConcurrentHashMap$Node
  13:       2023057       48M  java.util.ArrayList
  14:       1889233       45M  org.apache.ratis.server.raftlog.LogEntryHeader$1
  15:       1889233       45M  org.apache.ratis.server.raftlog.segmented.LogSegment$LogRecord
  16:       1082130       43M  org.rocksdb.Slice
  17:       2194983       35M  java.util.concurrent.atomic.AtomicBoolean
  18:       1082133       34M  org.rocksdb.ReadOptions
  19:       1082132       34M  org.rocksdb.RocksIterator
  20:         44604       29M  io.netty.util.internal.shaded.org.jctools.queues.MpscArrayQueue
  21:        519151       29M  org.apache.ratis.proto.RaftProtos$StateMachineLogEntryProto
  22:        410097       26M  com.google.protobuf.CodedInputStream$ArrayDecoder
  23:       1040706       24M  org.apache.ratis.thirdparty.com.google.protobuf.ByteString$LiteralByteString
  24:        485298       23M  java.util.HashMap
  25:        519154       20M  org.apache.ratis.proto.RaftProtos$MetadataProto
  26:        413188       16M  sun.nio.cs.UTF_8$Decoder
  27:        472992       15M  alluxio.master.metastore.caching.Cache$Entry
  28:        471806       15M  alluxio.security.authorization.AccessControlList
```

# Category 1: Primitive types and basic types

It's very hard to tell anything from looking at sizes alone. They can really be from anything. 

For example, String uses char[] in Java 8 and byte[] in Java 11. If there are a lot of Strings, you will see a large size in the String row and the corresponding primitive array row.

Another example, Long are very likely from the Collection<Long> or Map<Long, ...> that we use a lot in Alluxio. The Long is typically a file/block ID.

# Category 2: Finalizer

This one is very interesting. It does NOT look normal to have 3 million Finalizers in your JVM. 

This looks like the Finalizer thread in your JVM. 
```
Thread 3 (Finalizer):
  State: WAITING
  Blocked count: 104570
  Waited count: 23
  Waiting on java.lang.ref.ReferenceQueue$Lock@1b442707
  Stack:
    java.lang.Object.wait(Native Method)
    java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:144)
    java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:165)
    java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:216)
```
I can't really see what the problem is. Maybe there are some objects that take long to finalize, or there can simply be too many objects. Please dig deeper on this one. You really should take a heap dump and investigate what the Finalizer instances are. If you can upload your heap dump, I'd be happy to take a deeper look too.

I don't think any class in Alluxio relies on the `finalize()` method. Can't really guess where these are from.

# Category 3: Alluxio metadata storage

All these MutableInode, [Ljava.util.concurrent.ConcurrentHashMap$Node, ConcurrentHashMap$Node and almost all HashMap are from Alluxio's metadata storage. For example, the Maps are mapping the file/block IDs to their metadata. Although you are using RocksDB for metadata storage, Alluxio will internally keep some hot metadata on heap. The size looks normal to me.

# Category 4: RAFT(ratis) related classes

Alluxio embedded journal uses the RAFT protocol and ratis is the RAFT implementation we use. I do see many top-ranking classes with `org.apache.ratis`. At this moment I can't say if the RAFT-related logic can keep costing memory or even have a leak.

@jenoudet Before we can confirm if there's a "leak" with the HA part, could you double check if the RAFT related logic can cause the memory cost to keep growing? For example, could it be possible that log entries are not de-referenced so they cannot be GC-ed? Or under some circumstances the un-processed log entries keep queuing up?

# Category 5: others

There are many other classes in this histogram. However, they currently do not cost too much memory. At this moment I would guess the leak suspect is in the first 4 categories.

See my following comment about how to dig deeper with jmap.

**Comment by jiacheliu3** [comment]
In order to further investigate if there's a leak, we will need to analyze the heap multiple times in order to see what have been added. 

Since you say there's not much GC-events, I would start suspecting that all these objects are alive. In this case, you can rely on the `jmap -histo:live` which prints the live objects. If heap dumping is fast, you can retrieve multiple heap dumps too. Whichever way you choose, we are going to compare the jmap outputs or the heap dumps, in order to see what live objects are growing. You may need at least 3 of such snapshots.

By looking at the snapshots side by side, we should be able to further zoom into the leak suspects. If we can filter out 2 or 3 suspects, we can further reason about the leak scenario and test potential fixes. I'll be happy to help more(like schedule a screen-sharing meeting).

**Note that both `jmap -histo:live` and heap dump both trigger a full GC!**

**Comment by uniqueZt** [comment]
i also encounter the problem ，i use rocksdb store meta，but the old gen used always increase ，may be some data cannot be gc until master switch

**Comment by uniqueZt** [comment]
i think the inode size was limited，the memory should not always increase

**Comment by jiacheliu3** [comment]
@uniqueZt I think your understanding is mostly right. When there's no user operations, there are still some heartbeating checks in the master like `ReplicationChecker` and there's worker-master heartbeats which incurs RPC. But I don't think those objects will go to the old gen.

I will recommend performing a similar check by capturing multiple `jmap` or heap dumps so we better understand what are put into the old gen.

**Comment by jingcheng88** [comment]
@jiacheliu3
jmap -histo:live 
[jmap.txt](https://github.com/Alluxio/alluxio/files/7788315/jmap.txt)

I used offical docker image of community edition.
The memory would not leak when i added UseConcMarkSweepGC in JVM option. But it would crash with SIGSEGV.
Alluxio works well when i downgrade to version 2.5.0-3.

**Comment by jiacheliu3** [comment]
@jingcheng88 When you see SIGSEGV, that's probably due to RocksDB. You can find a file that looks like `hs_err.pid` that is essentially your core dump. You will find the thread stacktrace that causes the SIGSEGV. I'm almost certain you will see RocksDB. In that case, please create a separate issue to track the RocksDB crash, we will work on resolution there.

On using CMS vs the default ParallelOld, they are irrelevant to "leak". They just GC in different ways. If an object can be garbage collected, it will. Maybe you think there's a leak when you see OOM, and think there's no leak when you don't see OOM. That is not true. CMS may be able to GC faster. And ParallelOld may be slow so your memory runs out before it can GC enough. Does that match your case?

For your `jmap` output, yes I see difference from your last jmap output (without `:live`). For example I see `[I` and `[B` taking much less space. That is because in your JVM, the `int[]` and `byte[]` are mostly from temporary variables. They are not live any more and will be collected on the next GC.

In order to further check what's wrong, you should do `jmap -histo:live` multiple times, and compare the output to see the difference. If you see the space allocation in some objects keep growing, that means they are the top suspect for your "leak"!

**Comment by jiacheliu3** [comment]
@jingcheng88 could you find us on Slack so we can iterate faster? 
https://join.slack.com/t/alluxio-community/shared_invite/zt-ffbj8sli-wQlgpP0_Cr~4~kWmqcnnHQ

**Comment by jenoudet** [comment]
@jiacheliu3 After reviewing our HA Raft code I did not find anywhere that is susceptible to leak memory.

**Comment by jiacheliu3** [comment]
Hi @jingcheng88 We don't see more updates under this ticket and if the investigation does not go any further, we are going to close this ticket.

You can try:

1. There is a similar issue in https://github.com/Alluxio/alluxio/issues/14729 which turned out not to be a leak but related to GC tuning. @jingcheng88 Can you check your GC parameters and see that's relevant?

2. You may try later versions like 2.7.3. The 2.8.0 version will also be out soon, and that is supposed to include many memory optimizations.
